edb/pgsql/compiler/dml.py

```diff
@@ -891,6 +917,52 @@ def compile_insert_else_body(
     return anti_cte_iterator
 
 
+def compile_insert_else_body_failure_check(
+        on_conflict: irast.OnConflictClause,
+        *,
+        ctx: context.CompilerContextLevel) -> None:
+    else_fail = on_conflict.else_fail
+    if not else_fail:
+        return
+    # Copy the type rels from the possibly conflicting earlier DML
+    # into the None overlays so it gets picked up.
+    # TODO: Try to *only* get the overlay, not the base type.
+    ctx.type_rel_overlays = ctx.type_rel_overlays.copy()
+    ctx.type_rel_overlays[None].update(
+        ctx.type_rel_overlays[else_fail])
+    ctx.ptr_rel_overlays = ctx.ptr_rel_overlays.copy()
+    ctx.ptr_rel_overlays[None].update(
+        ctx.ptr_rel_overlays[else_fail])
+
+    assert on_conflict.constraint
+    cid = on_conflict.constraint.id
+    maybe_raise = pgast.FuncCall(
+        name=('edgedb', 'raise'),
+        args=[
+            pgast.TypeCast(
+                arg=pgast.NullConstant(),
+                type_name=pgast.TypeName(name=('text',))),
+            pgast.StringConstant(val='exclusion_violation'),
+            pgast.NamedFuncArg(
+                name='msg',
+                val=pgast.StringConstant(
+                    val=(
+                        f'duplicate key value violates unique '
+                        f'constraint "{cid};schemaconstr"'
+                    )
+                ),
+            ),
+            pgast.NamedFuncArg(
+                name='constraint',
+                val=pgast.StringConstant(val=f"{cid};schemaconstr")
```

Comment:
Need to pull this into `pgsql.common` to avoid hardcoding PostgreSQL object naming specifics in multiple places.